.circleci/config.yml

```diff
@@ -175,64 +160,113 @@ jobs:
 
-  deploy:
+  build-tar-and-docker-image:
     executor:
       name: elixir-builder
       env: prod
-
     steps:
       - checkout
       - setup_remote_docker:
           docker_layer_caching: true
       - restore_cache:
           keys:
-            - release-cache-{{ .Environment.CACHE_VERSION }}-{{ .Branch }}-{{ .Revision }}
+            - mix-cache-{{ .Environment.CACHE_VERSION }}-{{ .Branch }}-{{ .Revision }}
       - run:
-          name: Generate image tag
-          command: |
-             set -x
-             if [ -n "$CIRCLE_TAG" ]; then
-               DOCKER_TAG="$CIRCLE_TAG"
-             elif [ "$CIRCLE_BRANCH" == "master" ]; then
-               DOCKER_TAG="${CIRCLE_SHA1:0:7}"
-             else
-               DOCKER_TAG=`echo $CIRCLE_BRANCH | sed 's,/,_,g'`
-             fi
-             env | sort
-             echo "export DOCKER_TAG=$DOCKER_TAG" >> $BASH_ENV
-             echo "export DOCKER_IMAGE=mongoose-push" >> $BASH_ENV
-
+          name: Mix build release
+          command: mix do certs.dev, distillery.release
+      - save_cache:
+          key: certs-cache
+          paths:
+            - "priv"
+      - run:
+          name : Build tar archive
+          command: tar -czf mongoose_push.tar.gz -C _build/${MIX_ENV}/rel/mongoose_push .
+      - run: *DOCKER_TAGS
       - run:
           name: Login to DockerHub
           command: docker login -u "${DOCKERHUB_USER}" -p "${DOCKERHUB_PASS}"
-
       - run:
           name: Build docker image
-          command: docker build -f Dockerfile.release --build-arg MIX_ENV=prod -t $DOCKER_IMAGE:$DOCKER_TAG .
+          command: docker build -f Dockerfile.release --build-arg MIX_ENV=prod -t $DOCKERHUB_REPOSITORY/$DOCKER_IMAGE:$DOCKER_TAG .
+      - run:
+          name: Push docker image but do not tag it as the latest
+          command: docker push $DOCKERHUB_REPOSITORY/$DOCKER_IMAGE:$DOCKER_TAG
+
+  integration-tests:
+    machine:
+        image: ubuntu-1604:201903-01
+    steps:
+      - checkout
+      - run:
+          name: Install elixir
+          command: |
+              wget https://packages.erlang-solutions.com/erlang-solutions_1.0_all.deb && \
+              sudo dpkg -i erlang-solutions_1.0_all.deb && \
+              sudo apt-get update && \
+              sudo apt-get install -y esl-erlang=1:21.0 && \
+              sudo apt-get install -y elixir=1.8.0-1
+              wget https://github.com/jwilder/dockerize/releases/download/v0.6.1/dockerize-linux-amd64-v0.6.1.tar.gz && \
+              sudo chmod +x dockerize-linux-amd64-v0.6.1.tar.gz && \
+              sudo tar -C /usr/local/bin -xzvf dockerize-linux-amd64-v0.6.1.tar.gz && \
+              sudo chmod +x /usr/local/bin/dockerize && \
+              sudo rm dockerize-linux-amd64-v0.6.1.tar.gz
+
+      - restore_cache:
+          keys:
+              - mix-cache-{{ .Environment.CACHE_VERSION }}-{{ .Branch }}-{{ .Revision }}
+      - restore_cache:
+          keys:
+              - certs-cache
```

Comment:
We really should make this branch specific and add `{{ .Environment.CACHE_VERSION }}` because we may have conflicts between branches because of that.